### ISSUE TITLE
Do we need igraph_adjlist_init_from_inclist()?

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2264]
`igraph_adjlist_init_from_inclist()` is a somewhat recently and perhaps a bit hastily added function. This issue is to discuss whether it is a good idea to have it, or if it's better to remove it or replace it with something else.

 - It was added in commit 48bd7280371cdaf0a44c07c40f25f5aff989f7f9
 - It is only used unconditionally in `igraph_spanner()`

@ntamas, I think you added it when handling the PR for `igraph_spanner()`? Do you remember your reasoning for it?

The purpose it is used for in that function is to create both an inclist and a _matching_ adjlist. Questions to consider:

 - Could one simply produce an adjlist and an inclist separately?
    * Would that be slower?
    * Would that still guarantee that the adjlist and inclist match each other? As I recall, this was the motivation for introducing this function.
 - If the goal is to produce matching inclist / adjlist pairs, would it be better to have a function that creates both at the same time?

As discussed in #2263, many parts of igraph rely on adjlists being sorted. Therefore, there is some danger in having adjlist constructors that potentially produce non-sorted adjlists, especially in situations where it may feel natural to replace an `adjlist_init()` with an `adjlist_init_from_inclist()`.  Perhaps it is presently true that first using `inclist_init()`, then using `adjlist_init_from_inclist()` on the result yields a sorted adjlist. But relying on this fact is basically the same as relying on `igraph_inclist()` and `igraph_adjlist()` producing matching pairs when called with matching arguments. I suspect that we agree that we don't want to rely on such a property, as it can very easily be broken inadvertently.

For this reason, my suggestion is to replace `igraph_adjlist_init_from_inclist()` by a function that simultaneously initializes both an inclist and adjlist, and guarantees that (1) they match each other (2) the adjlist is sorted.  This is the cleanest approach, and likely the fastest one as well. It suits the present use case, and it is likely to suite most future use cases.

My suggestion is to introduce such a function now, and deprecate the public version of `igraph_adjlist_init_from_inclist()`. It does feel a bit unpleasant to already deprecate a recently introduced function, but such is the journey towards a high quality stable API.

**Comment by szhorvat** [comment]
If I recall correctly, @vtraag may have suggested introducing an "incadjlist" data structure that keeps both incident edges and adjacent vertices at the same time? Perhaps because we suspected `IGRAPH_OTHER` to be a source of slowdown? But as I remember, @GroteGnoom's benchmarking effort didn't quite confirm this suspicion (See #1823)?

**Comment by ntamas** [comment]
>  I think you added it when handling the PR for igraph_spanner()? Do you remember your reasoning for it?

Yes. Originally it was inlined into the `igraph_spanner()` function and it clearly represents an operation that is conceptually separate from creating a spanner graph, so it made sense to move it to its own function (and still does). Whether it should have been public or not is something that we could have debated back then, but it was released officially in 0.10 and I think it's not polite to our users to remove a function that was just added a minor version ago.

On the other hand, if vectors in an adjacency list are supposed to be sorted, then it would currently be the job of `igraph_adjlist_init_from_inclist()` to ensure this property, so yes, that should be fixed if this property does not hold yet.

**Comment by ntamas** [comment]
> if vectors in an adjacency list are supposed to be sorted, then it would currently be the job of `igraph_adjlist_init_from_inclist()` to ensure this property, so yes, that should be fixed if this property does not hold yet.

Hmm, okay, so there are two invariants that we could strive to keep:

1. Vectors in an adjacency list should be sorted. Right now this is not guaranteed.
2. When creating an adjacency list from an incidence list, the i-th element in each vector of the list should be the endpoint of the (same) i-th element in the incidence list. Right now this is what we guarantee.

If we wanted 1), then all it takes is adding an `igraph_vector_int_sort()` at the appropriate place. But then this breaks 2) if the original incidence list was shuffled in the wrong way. How shall we proceed here?

**Comment by ntamas** [comment]
I believe that both use-cases have some merit. Most constructors of `igraph_adjlist_t` return vertex IDs in ascending order, but this is not guaranteed or explicitly documented anywhere. Some igraph algorithms (like cocitation or bibliographic coupling) rely on the adjacency lists being sorted, while others do not. On the other hand, there are algorithms like `igraph_spanner()` and (most likely) `igraph_simple_cycles()` use an adjacency and an incidence list in parallel and do not care whether they are sorted as long as the two data structures match each other (i.e. edges appear in the same order in both data structures).

Internally, `igraph_adjlist_t` simply calls `igraph_neighbors()` on all vertices, and `igraph_inclist_t` simply calls `igraph_incident()`. This also holds for the lazy variants (but the calls are deferred). Whatever invariants these functions have _right now_ depend on the invariants provided by `igraph_neighbors()` and `igraph_incident()`, but this is an implementation detail.

So, all in all, I would not remove `igraph_adjlist_init_from_inclist()` but document that `igraph_adjlist_init()` and `igraph_adjlist_init_complementer()` do not guarantee that the adjacency lists are sorted, although this is the case right now. We reserve the right to change this in the future and if the caller relies on the vectors being sorted, then `igraph_adjlist_sort()` should be called explicitly. (We can skip this internally in our own algorithms as long as we do not change the implementation).

**Comment by ntamas** [comment]
Documentation added in fe918a8f1
